src/style/style_layer/fill_extrusion_style_layer.js

```diff
@@ -60,4 +72,143 @@ class FillExtrusionStyleLayer extends StyleLayer {
     }
 }
 
+function dot(a, b) {
+    return a.x * b.x + a.y * b.y;
+}
+
+function getIntersectionDistance(projectedQueryGeometry: Array<Array<Point>>, projectedFace: Array<Point>) {
+
+    if (projectedQueryGeometry.length === 1 && projectedQueryGeometry[0].length === 1) {
+        // For point queries calculate the z at which the point intersects the face
+        // using barycentric coordinates.
+        const a = projectedFace[0];
+        const b = projectedFace[1];
+        const c = projectedFace[3];
+        const p = projectedQueryGeometry[0][0];
+
+        const ab = b.sub(a);
+        const ac = c.sub(a);
+        const ap = p.sub(a);
+
+        const dotABAB = dot(ab, ab);
```

Comment:
This looked like magic to me at first, but I guess it's not actually that complicated:

- Find the barycentric coordinates of the projected point within the first triangle of the face, using _only_ the xy plane. It doesn't matter if the point is outside the first triangle because all the triangles in the face are in the same plane.
- Use the barycentric weighting along with the original triangle z coordinates to get the point of intersection.

The "barycentric" comment above was enough to point me in the right direction, but some elaboration along these lines would make it easier to read.

Reply:
Perfect！Is there a IOS & Android version?

Reply:
@BrightBeacon this will be ported to mapbox-gl-native but there is no ETA